design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,287 @@
+* **OCP UI**: Needs to ensure DataSource is linked properly to a PVC in order to determine whether the boot image for a common template is available. 
+
+
+# Design
+
+![Golden Image Importer](golden-image-delivery-images/importer-design.png "Golden Image Importer Design")
+
+A **DataImportCron** object results in its corresponding controller polling for updates to a container image on a recurring interval. When updates to a container image are detected, the DataImportCron will automatically import the contents into a new PVC (using the standard CDI import flows) residing in a specific namespace with a unique name. The DataImportCron may need to implement a PVC garbage collection policy for removing outdated PVCs after an expiration period has been exceeded.
+
+A **DataSource** object references an import/clone source for a PVC. Functionally, it is similar to the 'source' field of a DataVolume. The DataSource’s spec is mutable, which has the side effect of allowing the DataSource to act as a link that will redirect to whatever source the link currently points to. In the future, the DataSource will act as a PVC Populator as well as a DataVolume source.
+
+**DataImportCron can manage a corresponding DataSource** to always point to the latest most up-to-date PVC that the cron has created. The DataImportCron should also be capable of adopting and taking ownership of an existing DataSource in order to statisfy the condition where an admin wishs to opt-in to using auto delivery/updates later on.
+
+
+## CDI: DataVolume DataSource Source Field
+
+The DataVolume object will add a new field capable of importing pvc content using a DataSource. 
+
+![DataVolume DataSource Field](golden-image-delivery-images/volume-reference-source.png "DataVolume DataSource Field")
+
+## SSP: Setup DataImportCrons for common templates
+
+Since the SSP operator manages installing common templates, this operator should also be responsible for setting up the DataImportCrons used to automatically source the disk images into the cluster that the common templates use.
+
+The location the DataImportCrons pull from will be configurable on the SSP operators Custom Resource.
+
+![SSP Disk Import Pipelines](golden-image-delivery-images/ssp-disk-import-pipelines.png "SSP Disk Import Pipelines")
+
+## Common templates: DataSources as DataVolume Source
+
+The common templates shipped with SSP need to have their DataVolumeTemplate sections transition to using a DataSource rather than the inlined PVC source field. This allows the DataImportCron to seamlessly update the underlying PVC the corresponding DataSource points to, which then ensures that the common templates are always using the most up-to-date disk images.
+
+## HCO: Integration with SSP Operator
+
+In order to leverage golden disk image auto delivery/updates at the HCO level, the HCO is responsible for configuring DataImportCrons on the SSP operator’s CR.
```

Comment:
Do we have also to set initial defaults or opinionated values?
And if so, how can we deprecate/remove image set in the past on future releases (e.g. we configure something pointing to quay.io/kubevirt/centos8 now and in 2 years we want to remove it supporting only quay.io/kubevirt/centos9)?
What is supposed to happen to this defaults/opinionated values when running in disconnected environments?

Reply:
> And if so, how can we deprecate/remove image set in the past on future releases (e.g. we configure something pointing to quay.io/kubevirt/centos8 now and in 2 years we want to remove it supporting only quay.io/kubevirt/centos9)?

I think the expectation is that if we stop supporting automated updates for something like centos8, that the HCO removes the DataImportCron for centos8 on the SSP CR. That should result in the SSP operator deleting the cron object, but we can consider leaving the DataSource/PVC related to the last import in the cluster if we want. 

So automated updates would then be disabled for centos8, but the last boot image that was imported would remain available.

> What is supposed to happen to this defaults/opinionated values when running in disconnected environments?

It's possible this feature of auto delivery/updating boot images doesn't make sense for disconnected environments at all.

Reply:
> Do we have also to set initial defaults or opinionated values?

can you provide some more information on this? I'm not 100% sure what is being asked.

Reply:
I think it makes sense for disconnected environments. We talked about this before, they would just point the cron to their internal registry instead and have to take care of pushing the containerDisks according to their own process.

Reply:
>I think it makes sense for disconnected environments. We talked about this before, they would just point the cron to their internal registry instead and have to take care of pushing the containerDisks according to their own process. 

who is `they`. For the MVP, no one has control over the cron job except the HCO. Someone could create their own cron's outside of HCO and SSP control, but it wouldn't be the crons that the admin opts into when enabling automated image delivery/update

@tiraboschi for the introduction of this initial feature, is it reasonable that we don't target disconnected environments?

Reply:
Fine, let's specify it in the proposal as a reference for the future.

Reply:
alrighty, i updated the document and tried to briefly capture this in the `HCO Disconnected Environments` section.